index_impl.go

```diff
@@ -483,15 +485,24 @@ func (i *indexImpl) SearchInContext(ctx context.Context, req *SearchRequest) (sr
 		collector.SetFacetsBuilder(facetsBuilder)
 	}
 
-	if memCb := ctx.Value(SearchMemCheckCallbackKey); memCb != nil {
-		if memCbFn, ok := memCb.(SearchMemCheckCallbackFn); ok {
-			err = memCbFn(memNeededForSearch(req, searcher, collector))
+	memNeeded := memNeededForSearch(req, searcher, collector)
+	if cb := ctx.Value(SearchQueryStartCallbackKey); cb != nil {
+		if cbF, ok := cb.(SearchQueryStartCallbackFn); ok {
+			err = cbF(memNeeded)
 		}
 	}
 	if err != nil {
 		return nil, err
 	}
 
+	defer func() {
+		if cb := ctx.Value(SearchQueryEndCallbackKey); cb != nil {
+			if cbF, ok := cb.(SearchQueryEndCallbackFn); ok {
+				_ = cbF(memNeeded)
```

Comment:
maybe we could reverse these.  meaning, only if the 2 conditions are satisfied do we bother defering anything at all.  then we can just defer cbF and not an anonymous function.

Reply:
Hmm - ya I could do that, but with defer cbF(memNeeded), errCheck complains that we're not handling the response of the function.

Now a return response is not needed for the QueryEndCallback, but if I were to change the definition of the SearchQueryEndCallbackFn, I cannot use this switch-case in cbft:

        func (a *appHerder) onQueryEvent(event cbft.QueryEvent, size uint64) error {
	        switch event.Kind {
	        case cbft.EventQueryStart:
		        return a.onQueryStart(size)

	        case cbft.EventQueryEnd:
		        return a.onQueryEnd(size)

	        default:
		        return nil
	        }
        }

Is there a reason why you'd rather I not defer an anonymous function, or perhaps is this something that's ok too:

        if cb := ctx.Value(SearchQueryEndCallbackKey); cb != nil {
                if cbF, ok := ok.(SearchQueryEndCallbackFn); ok {
                        defer func() {
                                _ = cbF(memNeeded)
                        }
                }
        }

Reply:
Ya this is fine, there is no problem with the anonymous func, just wanted to avoid it if possible.

Reply:
My first instinct is that the query event stuff is over-engineered.  But, I guess I can kinda see it being used for query stats in the future, so let's go with it.

Reply:
Yep, my thoughts too.